devtools/src/arcs-bug-report.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {PolymerElement} from '../deps/@polymer/polymer/polymer-element.js';
+import {html} from '../deps/@polymer/polymer/lib/utils/html-tag.js';
+import {MessengerMixin} from './arcs-shared.js';
+import '../deps/jszip/dist/jszip.js';
+
+/**
+ * Saves/loads a bug report which contains debug messages sent to devtools.
+ */
+class ArcsBugReport extends MessengerMixin(PolymerElement) {
+  static get template() {
+    return html`
+    <style include="shared-styles">
+      :host {
+        display: flex;
+        flex-direction: column;
+      }
+    </style>
+    <div class="wrapper">
+      <input type="file" id="fileElem" style="display:none" on-change="_loadBugReport">
+      <button id="fileSelect">Load Bug Report</button>
+      <button type="button" on-click="_saveBugReport">Save Bug Report</button>
+    </div>`;
+  }
+
+  static get is() { return 'arcs-bug-report'; }
+
+  constructor() {
+    super();
+    this.messages = [];
+  }
+
+  ready() {
+    super.ready();
+    const fileElem = this.$.fileElem;
+    this.$.fileSelect.addEventListener("click", function (e) {
+      if (fileElem) {
+        fileElem.click();
+      }
+    }, false);
+  }
+
+  onRawMessageBundle(messages) {
+    this.messages = this.messages.concat(messages);
+  }
+
+  _saveBugReport() {
+    var zip = new JSZip();
+    var file = zip.file("bugreport.txt", JSON.stringify(this.messages));
+    file.generateAsync({ type: "blob" })
+      .then(function (blob) {
+        const a = document.createElement('a');
+        a.download = `bugreport.zip`;
+        a.href = window.URL.createObjectURL(blob);
+        a.click();
+      });
+  }
+
+  _loadBugReport(event) {
+    let files = event.target.files;
+    for (var i = 0; i < files.length; i++) {
+      JSZip.loadAsync(files[i]).then(
+        zip => this._decompress(zip),
+        e => alert("Error reading " + f.name + ": " + e.message));
+    }
+  }
+
+  _decompress(zip) {
+    zip.forEach(
+      relativePath => zip.file(relativePath).async("text").then(
+        txt => this._handleLog(txt)));
+  }
+
+  _handleLog(txt) {
+    let messages = JSON.parse(txt);
+    this.emitFilteredMessages(messages);
```

Comment:
You may have noticed that it doesn't fully replicate the state the Arcs Explorer was in. (e.g. all the arcs are shown together in the overview, you cannot select between arcs etc.

This is because you are replaying the messages as these "filtered messages", which should only pertain to a single arc. arcs-selector is the component that is responsible for switching between arcs and replaying the messages, but it itself listens to messages which are _not_ filtered.

To sum up, you should not emit the messages as filtered messages, you should emit them as they are and let the webapp deal with them as it would. The easiest way to do it would be:

`document.dispatchEvent(new CustomEvent('raw-messages', {detail: messages}));`

A slightly nicer way would be to have a concept of "replayed connection" in devtools.js, where all the connection types are listed. I'll let you explore that if you're interested.